### PR TITLE
Single file multi-threading (read-only)

### PR DESCRIPTION
**Pull request by ejmeitz**
This PR aims to add support for multi-threaded read-only access to JLD2 files. I created a new global tracker `OPEN_PARALLEL_FILES` which keeps track of the files flagged as open for multi-threaded access. The flag for multi-threaded access was added to the `jldopen` function as `parallel_read` with a default of false. If the user changes this flag to true the program will check that the file is opened with in mode `r` and that it is not already present in the `OPEN_FILES` list. If these checks are satisfied `jldopen` proceeds as normal. The `OPEN_FILES_LOCK` is not invoked when `parallel_read` is true.

This is in reference to #403. Still need to add tests, but the existing tests pass locally.

Things I am unsure of:
- Do I need to verify compression types are passed the same each time? (other flags as well)
- How to write tests that guarantee parallel reads (e.g. could just get lucky with timing and one finishes before the other). I dont seem to be able to spawn threads in a testset.
- Why the BigInt tests fails on two of the architectures in the CI (passes locally for me)

### CODE REVIEW

File: src/JLD2.jl [hunk 1]
```diff
@@ -334,23 +344,38 @@ function jldopen(fname::AbstractString, wr::Bool, create::Bool, truncate::Bool,
...
+                end
+            elseif parallel_read && haskey(OPEN_FILES, rname)
+                throw(ArgumentError("Tried to open file in a parallel context but it is open elsewhere in a serial context."))
+            elseif parallel_read && haskey(OPEN_PARALLEL_FILES, rname)
+                ref = OPEN_PARALLEL_FILES[rname]
+                f = ref.value
+                if !isnothing(f)
                     f = f::JLDFile{iotype}
                     f.n_times_opened += 1
                     return f
```
Comment (JonasIsensee) [inline src/JLD2.jl:359]: Sadly, I think, this does not solve the problem at all.

With this logic you still just end up with a single open file handle.
This should should return a new file handle each time.

**Review by JonasIsensee — CHANGES_REQUESTED** [review]
Hi

thank you for your contribution!

At this stage, this PR will not do what we want it to do.

I like the new keyword argument and that you check for the file to be in read-only mode.
However, the `JLDFile` object doesn not support parallel access.
To circumvent that, this PR must return a new `JLDFile` instance every time you do `jldopen(filename)`.

Possibly contrary to what I said earlier, maybe we don't need to keep track of `OPEN_PARALLEL_FILES`. Currently it can't hold multiple file handles of the same file anyway. The only advantage would be to throw an error when a user first opens a `parallel-read` file handle, then edits the file in regular mode and continues to read with the original file handle.

**Comment by ejmeitz** [comment]
Makes sense, that shouldn't be too hard to remedy.

File: src/JLD2.jl [hunk 2]
```diff
@@ -297,26 +297,36 @@ FallbackType(::Type{IOStream}) = nothing
...
     exists = ispath(fname)
 
-    lock(OPEN_FILES_LOCK)
+    # Can only open multiple in parallel if mode is "r"
+    if parallel_read && (wr, create, truncate)  != (false, false, false)
+        throw(ArgumentError("Cannot open file in a parallel context unless mode is \"r\""))
+    end
+
+    #Do not lock file if user specifies parallel_read
+    parallel_read && lock(OPEN_FILES_LOCK)
```
[the remark this answers is not in the document]
Comment (ejmeitz) [reply src/JLD2.jl:319]: My idea with the lock was to only use it if the users is reading the files normally. Do you think I still need some kind of lock on the file handles with the parallel access?

[the remark this answers is not in the document]
Comment (ejmeitz) [reply src/JLD2.jl]: I guess this is unnecessary if every parallel access gets a new file handle.

**Comment by ejmeitz** [comment]
Well the code definitely makes a new handle every time if `parallel_read` is true but I am still clueless as to how I can test this does what I think it does.

**Comment by JonasIsensee** [comment]
> Why the BigInt tests fails on two of the architectures in the CI (passes locally for me)
Don't worry about this one. It happens sporadically. I'm not sure why.

**Comment by JonasIsensee** [comment]
Thank you for making the changes!

That is fine - I don't think we can properly test this through CI. I'll have a final close look at the code and test it locally.

Looking at the code now, I don't think I like the "OPEN_PARALLEL_FILES" list. I think we're better off without it.

Also, you (probably accidentally) committed a manifest and a `.jld2` file.

**Comment by ejmeitz** [comment]
- I'll try and test it out as well. I have some of the banned use cases in tests but not actual parallel reads.
- I left the `OPEN_PARALLEL_FILES` to check that case you mentioned of someone trying to open the same file elsewhere with `parallel_read` set to false and then writing to that file.
- Yeah the manifest is from running the tests locally oops.

**Comment by JonasIsensee** [comment]
> * I'll try and test it out as well. I have some of the banned use cases in tests but not actual parallel reads.

:+1: 

>     * I left the `OPEN_PARALLEL_FILES` to check that case you mentioned of someone trying to open the same file elsewhere with `parallel_read` set to false and then writing to that file.

I understand and appreciate the intention. I'd also support it if it were that simple.
Imagine the following test-case.
```
fp = jldopen(testfile; parallel_read=true)
close(fp) # probably (already) fails because it can't be deleted from the OPEN_FILES Dict

fs = jldopen(testfile; parallel_read=false) # fails because OPEN_PARALLEL_FILES has an entry (that may be WeakRef(nothing))
```

Even harder to fix:
```
fp1 = jldopen(testfile; parallel_read=true)
fp2 = jldopen(testfile; parallel_read=true) # overwrites reference tot fp1
close(fp2)

# If you fix the case above, JLD2 would now still think that there is no open parallel file handle.
fs = jldopen(tesfile; parallel_read=false) # should fail according to current logic but doesn't  
```


Since this is quite a bit of work (Would need to keep a full list of ALL open file handles and also check whether they are still open) I would prefer removing this safety check and add a warning in the docs instead.

 
>     * Yeah the manifest is from running the tests locally oops. It won't let me delete the test.jld2 file I committed for some reason.

That's fine. I can fix that in the end.

**Comment by JonasIsensee** [comment]
Hi @ejmeitz,

thanks for helping out!
I had a few minutes to spare and took the liberty to finish this up real quick.

I relaxed the open conditions. (parallel and serial read is fine)
Will merge and release once tests pass.

**Review by JonasIsensee — APPROVED** [review]

**Comment by JonasIsensee** [comment]
@ZenanH 
The JLD2 part looks fine by itself.
You are reusing the iteration variable i a bunch of times. You could check that this is not accidentally messing things up in the rest of your code.

**Comment by ejmeitz** [comment]
I use this feature in my code and it works correctly, so I think JLD2 is fine.  I don't really understand your code but there could be a race condition in there somewhere. Its hard to know if the issue is JLD2 or elsewhere.

**Pull request merged**